### PR TITLE
Add CVE applicability source code scanning

### PR DESCRIPTION
**Pull request by Or-Geva**
- [x] All [tests](https://github.com/jfrog/jfrog-vscode-extension#building-and-testing-the-sources) passed. If this feature is not already covered by the tests, I added new tests.
- [x] I used `npm run format` for formatting the code before submitting the pull request.
-----
image (7)

Depending on:

- https://github.com/jfrog/jfrog-client-js/pull/59

### CODE REVIEW

File: src/main/treeDataProviders/sourceCodeTree/sourceCodeFileTreeNode.ts [hunk 1]
```diff
@@ -1,80 +1,64 @@
...
-//             this.description = _filePath;
-//         }
-//     }
+    constructor(
+        private _filePath: string,
+        private _children: SourceCodeCveTreeNode[],
+        private _parent?: SourceCodeRootTreeNode,
+        collapsibleState?: vscode.TreeItemCollapsibleState
+    ) {
+        super(_filePath.substring(_filePath.lastIndexOf(path.sep) + 1), collapsibleState ?? vscode.TreeItemCollapsibleState.Expanded);
```
Comment (yahavi) [inline src/main/treeDataProviders/sourceCodeTree/sourceCodeFileTreeNode.ts:17]: It was not trivial to me to understand what this substring does. Does it extract the directory name or the file name?
Also, I think I saw this repeated in SourceCodeRootTreeNode. Let's export this functionality to a utility.

File: src/main/treeDataProviders/sourceCodeTree/sourceCodeFileTreeNode.ts [hunk 2]
```diff
@@ -1,80 +1,64 @@
...
-//         const node: SourceCodeFileTreeNode = new SourceCodeFileTreeNode(
-//             'No vulnerabilities found',
-//             [],
-//             undefined,
-//             vscode.TreeItemCollapsibleState.None
-//         );
-//         node._topSeverity = Severity.Normal;
-//         return node;
-//     }
+    public static createFailedScan(): SourceCodeFileTreeNode {
```
Comment (yahavi) [inline src/main/treeDataProviders/sourceCodeTree/sourceCodeFileTreeNode.ts:26]:
```suggestion
    public static createFailedScanNode(): SourceCodeFileTreeNode {
```

File: src/main/treeDataProviders/sourceCodeTree/sourceCodeRootTreeNode.ts [hunk 3]
```diff
@@ -1,97 +1,88 @@
...
-//     }
-// }
+import * as vscode from 'vscode';
+import * as path from 'path';
+import { PackageType } from '../../types/projectType';
+import { SourceCodeCveTreeNode } from './sourceCodeCveNode';
+import { SourceCodeFileTreeNode } from './sourceCodeFileTreeNode';
+import { CveApplicabilityRoot } from './cveApplicabilityRoot';
+
+// Represent a root project node containing vulnerable files, in the CVE Applicability view.
```
Comment (yahavi) [inline src/main/treeDataProviders/sourceCodeTree/sourceCodeRootTreeNode.ts:8]: Not Go
```suggestion
/**
* Represent a root project node containing vulnerable files, in the CVE Applicability view.
*/
```

File: src/main/treeDataProviders/sourceCodeTree/sourceCodeRootTreeNode.ts [hunk 4]
```diff
@@ -1,97 +1,88 @@
...
+export class SourceCodeRootTreeNode extends vscode.TreeItem {
+    private _notApplicableCves: Set<string> = new Set<string>();
+    private _children: SourceCodeFileTreeNode[] = [];
+    // Map of cve -> cve tree node.
```
Comment (yahavi) [inline src/main/treeDataProviders/sourceCodeTree/sourceCodeRootTreeNode.ts:12]: This is obvious ;)
```suggestion
```

File: src/main/treeDataProviders/sourceCodeTree/sourceCodeTreeDataProvider.ts [hunk 5]
```diff
@@ -1,281 +1,266 @@
...
+    }
 
-//     /**
-//      * Return a map of 'workspace' -> 'CveApplicabilityTree'
-//      */
-//     public getScannedProjects(): Map<string, SourceCodeRootTreeNode> {
-//         return this.workspaceToTree;
-//     }
+    public isCveNotApplicable(workspace: string, cve: string): boolean {
+        return this.workspaceToTree.get(workspace)?.noApplicableCves.has(cve) || false;
```
Comment (yahavi) [inline src/main/treeDataProviders/sourceCodeTree/sourceCodeTreeDataProvider.ts:153]:
```suggestion
        return this.workspaceToTree.get(workspace)?.isCveNotApplicable(cve) || false;
```

File: src/main/treeDataProviders/sourceCodeTree/sourceCodeTreeDataProvider.ts [hunk 6]
```diff
@@ -1,281 +1,266 @@
...
-//     }
+    public isFileScanned(file: string): boolean {
+        return this.fileToNode.has(file) || false;
+    }
 
-//     public isCveNotApplicable(workspace: string, cve: string): boolean {
-//         return this.workspaceToTree.get(workspace)?.noApplicableCves.has(cve) || false;
-//     }
+    public isCveApplicable(workspace: string, cve: string): boolean {
+        return this.workspaceToTree.get(workspace)?.applicableCves.has(cve) || false;
```
Comment (yahavi) [inline src/main/treeDataProviders/sourceCodeTree/sourceCodeTreeDataProvider.ts:145]:
```suggestion
        return this.workspaceToTree.get(workspace)?.isCveApplicable(cve) || false;
```

File: src/main/treeDataProviders/sourceCodeTree/sourceCodeTreeDataProvider.ts [hunk 7]
```diff
@@ -1,281 +1,266 @@
...
+    }
 
-//     public getFileTreeNode(file: string): SourceCodeFileTreeNode | undefined {
-//         if (!this.isFileScanned(file)) {
-//             return undefined;
-//         }
-//         return this.fileToNode.get(file);
-//     }
+    public getCveApplicable(workspace: string, cve: string): SourceCodeCveTreeNode | undefined {
```
Comment (yahavi) [inline src/main/treeDataProviders/sourceCodeTree/sourceCodeTreeDataProvider.ts:148]:
```suggestion
    public getApplicableCve(workspace: string, cve: string): SourceCodeCveTreeNode | undefined {
```

File: src/main/treeDataProviders/sourceCodeTree/sourceCodeTreeDataProvider.ts [hunk 8]
```diff
@@ -1,281 +1,266 @@
...
+        this.workspaceToTree = new Map<string, SourceCodeRootTreeNode>();
+    }
+
+    public async refresh() {
+        this.clearCveTreeView();
+        const version: string | undefined = this._treesManager.sourceCodeTreeDataProvider._cveApplicabilityRunner.version();
+        if (version == undefined) {
+            return;
+        }
+        this._treesManager.logManager.logMessage("Running CVE Applicability  version '" + version.trim() + "'", 'INFO');
```
Comment (yahavi) [inline src/main/treeDataProviders/sourceCodeTree/sourceCodeTreeDataProvider.ts:262]:
```suggestion
        this._treesManager.logManager.logMessage("Running CVE Applicability version '" + version.trim() + "'", 'INFO');
```

File: src/main/diagnostics/applicabilityActionProvider.ts [hunk 9]
```diff
@@ -1,99 +1,103 @@
...
-//                 );
-//                 diagnostic.source = this.getSource();
-//                 diagnostic.code = node.cve;
-//                 diagnostics.push(diagnostic);
-//             }
-//         });
-//         this._diagnosticCollection.set(document.uri, diagnostics);
-//     }
+    /**
+     * Updates a diagnostics in a document based on CVEs  found in it.
```
Comment (yahavi) [inline src/main/diagnostics/applicabilityActionProvider.ts:60]:
```suggestion
     * Updates diagnostics in a document based on CVEs found in it.
```

File: src/main/diagnostics/applicabilityActionProvider.ts [hunk 10]
```diff
@@ -1,99 +1,103 @@
...
+        }
+        let diagnostics: vscode.Diagnostic[] = [];
+        scanResult.children.forEach(node => {
+            for (const nodeDetails of node.getNodeDetails()) {
+                const startPos: vscode.Position = new vscode.Position(nodeDetails.startLine - 1, nodeDetails.startColumn);
+                const endPosition: vscode.Position = new vscode.Position(nodeDetails.endLine - 1, nodeDetails.endColumn);
+                let diagnostic: vscode.Diagnostic = new vscode.Diagnostic(
+                    new vscode.Range(startPos, endPosition),
+                    nodeDetails.codeIssue,
+                    DiagnosticsUtils.toDiagnosticSeverity(node.severity ?? Severity.Critical)
```
Comment (yahavi) [inline src/main/diagnostics/applicabilityActionProvider.ts:75]:
```suggestion
                    DiagnosticsUtils.toDiagnosticSeverity(node.severity ?? Severity.Unknown)
```

File: src/main/treeDataProviders/sourceCodeTree/sourceCodeTreeDataProvider.ts [hunk 11]
```diff
@@ -1,281 +1,266 @@
...
-//     }
+    getTreeItem(
+        element: SourceCodeRootTreeNode | SourceCodeFileTreeNode | SourceCodeCveTreeNode | CveApplicabilityRoot
+    ): vscode.TreeItem | Thenable<vscode.TreeItem> {
+        if (element instanceof SourceCodeRootTreeNode) {
+            element.iconPath = PackageDescriptorUtils.getIcon(element.projectType);
+            return element;
+        }
+        if (element instanceof SourceCodeFileTreeNode) {
+            element.iconPath = SeverityUtils.getIcon(element.topSeverity !== undefined ? element.topSeverity : Severity.Critical);
```
Comment (yahavi) [inline src/main/treeDataProviders/sourceCodeTree/sourceCodeTreeDataProvider.ts:178]:
```suggestion
            element.iconPath = SeverityUtils.getIcon(element.topSeverity !== undefined ? element.topSeverity : Severity.Unknown);
```

File: src/main/treeDataProviders/sourceCodeTree/sourceCodeTreeDataProvider.ts [hunk 12]
```diff
@@ -1,281 +1,266 @@
...
+            return element;
+        }
+        if (element instanceof SourceCodeCveTreeNode) {
+            element.command = {
+                command: 'jfrog.source.code.scan.jumpToSource',
+                title: 'Jump To Code',
+                arguments: [element]
+            };
+            element.iconPath = SeverityUtils.getIcon(element.severity !== undefined ? element.severity : Severity.Critical);
```
Comment (yahavi) [inline src/main/treeDataProviders/sourceCodeTree/sourceCodeTreeDataProvider.ts:187]:
```suggestion
            element.iconPath = SeverityUtils.getIcon(element.severity !== undefined ? element.severity : Severity.Unknown);
```

File: src/main/types/applicabilityScanResults.ts [hunk 13]
```diff
@@ -1,21 +1,16 @@
...
- * It will be hidden until it is officially released.        *
- * ***********************************************************
- */
-// export interface ApplicabilityScanResults {
-//     results: Map<string, ApplicabilityScanResult[]>;
-//     scanners_ran: string[];
-// }
+export interface ApplicabilityScanResults {
+    results: Map<string, ApplicabilityScanResult[]>;
+    scanners_ran: string[];
```
Comment (yahavi) [inline src/main/types/applicabilityScanResults.ts:3]: What is scanners_ran? Please explain in a comment.

File: src/main/types/projectDetails.ts [hunk 14]
```diff
@@ -0,0 +1,56 @@
+import { ComponentDetails } from 'jfrog-client-js';
+import * as path from 'path';
+import { PackageType } from './projectType';
+import Set from 'typescript-collections/dist/lib/Set';
+
+export class ProjectDetails {
+    // dependencyGAV -> ComponentDetails
```
Comment (yahavi) [inline src/main/types/projectDetails.ts:7]:
```suggestion
```

File: src/main/types/projectDetails.ts [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+    private _dependencies: Set<ComponentDetails> = new Set<ComponentDetails>();
+    private _name: string;
+    constructor(private _path: string, private _type: PackageType) {
```
Comment (yahavi) [inline src/main/types/projectDetails.ts:10]:
```suggestion
    private _name: string;
    
    constructor(private _path: string, private _type: PackageType) {
```

File: src/main/utils/configuration.ts [hunk 16]
```diff
@@ -31,17 +31,13 @@ export class Configuration {
...
-     * ***********************************************************
-     */
-    // public static getRemoteArtifactory(): string {
-    //     return vscode.workspace
-    //         .getConfiguration('jfrog')
-    //         .get('jfrog.remote.artifactory', '')
-    //         .trim();
-    // }
+
+    public static getRemoteArtifactory(): string {
```
Comment (yahavi) [inline src/main/utils/configuration.ts:35]: This function looks broken for many reasons. Let's remove or fix it.

File: src/main/utils/iconsPaths.ts [hunk 17]
```diff
@@ -27,16 +28,24 @@ export class IconsPaths {
...
+
+export class PackageDescriptorUtils {
+    public static getIcon(packageType: PackageType): string {
+        switch (packageType) {
+            case PackageType.PYTHON:
+                return IconsPaths.PYTHON;
+            case PackageType.NPM:
+                return IconsPaths.NPM;
+            default:
+                return IconsPaths.NPM;
```
Comment (yahavi) [inline src/main/utils/iconsPaths.ts:48]:
```suggestion
                return undefined;
```

File: src/main/utils/resource.ts [hunk 18]
```diff
@@ -1,112 +1,112 @@
...
-//         'https://releases.jfrog.io/artifactory',
-//         '',
-//         '',
-//         '',
-//         ''
-//     );
+// Resource classes represent generic instances of external artifact (e.g. binary).
+export class Resource {
+    // You should not change this variable. This may only be done for testing purposes.
+    public static MILLISECONDS_IN_HOUR: number = 3600000;
```
Comment (yahavi) [inline src/main/utils/resource.ts:13]: Please make sure to do that in all contants
```suggestion
    public static readonly MILLISECONDS_IN_HOUR: number = 3600000;
```

File: src/main/utils/resource.ts [hunk 19]
```diff
@@ -1,112 +1,112 @@
...
 
-//             readStream.on('error', err => reject(err));
-//             writeStream.on('error', err => reject(err));
+    // Update the resource to the latest released
+    public async update(withExecPrem: boolean) {
+        const updateAvailable: boolean = await this.isUpdateAvailable();
+        if (!updateAvailable) {
+            return;
+        }
+        this._logManager.logMessage('Downloading new update from ' + this.downloadTarget, 'DEBUG');
```
Comment (yahavi) [inline src/main/utils/resource.ts:58]: I think the user has the right to know what are we downloading for him

File: src/main/utils/resource.ts [hunk 20]
```diff
@@ -1,112 +1,112 @@
...
+        if (this.isUpdateStarted()) {
+            if (this.isUpdateStuck() === false) {
```
Comment (yahavi) [inline src/main/utils/resource.ts:60]:
```suggestion
            if (!this.isUpdateStuck()) {
```

File: src/main/utils/resource.ts [hunk 21]
```diff
@@ -1,112 +1,112 @@
...
+        try {
+            await this._connectionManager
+                .artifactory()
+                .download()
+                .downloadArtifactToFile(this.downloadSource, this.downloadTarget);
+            if (withExecPrem) {
+                fs.chmodSync(this.downloadTarget, '755');
+            }
+            fs.copyFileSync(this.downloadTarget, this.path);
+            this._logManager.logMessage('Update resource was successfully upgraded for ' + this.downloadSource, 'DEBUG');
```
Comment (yahavi) [inline src/main/utils/resource.ts:75]: "Update resource was successfully upgraded"?
Let's rephrase this sentence and replace "resource" with the binary name.

File: src/main/utils/resource.ts [hunk 22]
```diff
@@ -1,112 +1,112 @@
...
-//                     .downloadArtifactToFile(this.resourceRemotePath, path.join(this.TMP_DOWNLOAD_PATH, this.resourceName));
-//             }
+    constructor(
+        private _homeDir: string,
+        private downloadSource: string,
+        resourceName: string,
+        private _logManager: LogManager,
+        connectionManager?: JfrogClient
+    ) {
+        this.downloadDir = path.join(_homeDir, 'download');
```
Comment (yahavi) [inline src/main/utils/resource.ts:39]: I'm not sure that using a **temporary** dir under the user home is a good idea. The system's temp folder is meant for this.
Also - if the user closes his vscode during download, the next time he opens vscode may cause freeze for an hour. Using a temp folder in the system temp path should resolve this.

File: src/main/utils/resource.ts [hunk 23]
```diff
@@ -1,112 +1,112 @@
...
-//             const readStream: fs.ReadStream = fs.createReadStream(oldPath);
-//             const writeStream: fs.WriteStream = fs.createWriteStream(newPath);
+    public get homeDir(): string {
+        return this._homeDir;
+    }
 
-//             readStream.on('error', err => reject(err));
-//             writeStream.on('error', err => reject(err));
+    // Update the resource to the latest released
+    public async update(withExecPrem: boolean) {
```
Comment (yahavi) [inline src/main/utils/resource.ts:53]: Looks always true.
```suggestion
    public async update() {
```

File: src/main/utils/resource.ts [hunk 24]
```diff
@@ -1,112 +1,112 @@
...
+            }
+            fs.rmSync(this.downloadDir, { recursive: true });
+        }
+        fs.mkdirSync(this.downloadDir, { recursive: true });
+        try {
+            await this._connectionManager
+                .artifactory()
+                .download()
+                .downloadArtifactToFile(this.downloadSource, this.downloadTarget);
+            if (withExecPrem) {
```
Comment (yahavi) [inline src/main/utils/resource.ts:71]: I'd replace it to a check whether OS != Windows

File: src/main/utils/resource.ts [hunk 25]
```diff
@@ -1,112 +1,112 @@
...
+            this._logManager.logMessage('Update resource was successfully upgraded for ' + this.downloadSource, 'DEBUG');
+        } finally {
+            fs.rmSync(this.downloadDir, { recursive: true, force: true });
+        }
+    }
 
-//             writeStream.on('close', function() {
-//                 resolve(true);
-//             });
+    public async isUpdateAvailable(): Promise<boolean> {
```
Comment (yahavi) [inline src/main/utils/resource.ts:81]:
```suggestion
    public async shouldUpdate(): Promise<boolean> {
```

File: src/main/utils/resource.ts [hunk 26]
```diff
@@ -1,112 +1,112 @@
...
+        if (!fs.existsSync(this.path)) {
+            return true;
+        }
+        let checksumResult: IChecksumResult = { sha256: '', sha1: '', md5: '' };
```
Comment (yahavi) [inline src/main/utils/resource.ts:85]: I think that sha1 + md5 are not really needed.
```suggestion
        let checksumResult: IChecksumResult = { sha256: '' };
```

File: src/test/tests/cveApplicabilityTree.test.ts [hunk 27]
```diff
@@ -0,0 +1,182 @@
...
+        let sourceCodeRootTreeNode: SourceCodeRootTreeNode = new SourceCodeRootTreeNode(
+            path.join('path', 'to', 'my-project'),
+            PackageType.NPM,
+            'my-project'
+        );
+        sourceCodeRootTreeNode.addChild(sourceCodeFileTreeNode);
+        assert.strictEqual(sourceCodeRootTreeNode.label, 'my-project');
+        assert.strictEqual(sourceCodeRootTreeNode.description, path.join('path', 'to', 'my-project'));
+        assert.strictEqual(sourceCodeRootTreeNode.isCveApplicable(cve1), false);
+        assert.strictEqual(sourceCodeRootTreeNode.isCveNotApplicable(cve1), false);
```
Comment (yahavi) [inline src/test/tests/cveApplicabilityTree.test.ts:100]:
```suggestion
        assert.isFalse(sourceCodeRootTreeNode.isCveApplicable(cve1));
        assert.isFalse(sourceCodeRootTreeNode.isCveNotApplicable(cve1));
```

File: src/test/tests/cveApplicabilityTree.test.ts [hunk 28]
```diff
@@ -0,0 +1,182 @@
...
+
+        sourceCodeRootTreeNode.noApplicableCves = new Set<string>([cve1]);
+        sourceCodeRootTreeNode.applicableCves = new Map<string, SourceCodeCveTreeNode>([[cve2, node2]]);
+
+        assert.strictEqual(sourceCodeRootTreeNode.isCveNotApplicable(cve1), true);
+        assert.strictEqual(sourceCodeRootTreeNode.isCveNotApplicable(cve2), false);
+
+        assert.strictEqual(sourceCodeRootTreeNode.isCveApplicable(cve2), true);
+        assert.strictEqual(sourceCodeRootTreeNode.isCveApplicable(cve1), false);
```
Comment (yahavi) [inline src/test/tests/cveApplicabilityTree.test.ts:109]:
```suggestion
        assert.isTrue(sourceCodeRootTreeNode.isCveNotApplicable(cve1));
        assert.isFalse(sourceCodeRootTreeNode.isCveNotApplicable(cve2));

        assert.isTrue(sourceCodeRootTreeNode.isCveApplicable(cve2));
        assert.isFalse(sourceCodeRootTreeNode.isCveApplicable(cve1));
```

File: src/test/tests/cveApplicabilityTree.test.ts [hunk 29]
```diff
@@ -0,0 +1,182 @@
...
+        workspaceFolders = [
+            {
+                uri: tmpDir,
+                name: '',
+                index: 0
+            } as vscode.WorkspaceFolder
+        ];
+    });
+
+    it('Source Code Cve Tree Node', () => {
```
Comment (yahavi) [inline src/test/tests/cveApplicabilityTree.test.ts:51]: All "it" should be lowercase. Let's make this change in all other unit tests in this file.
```suggestion
    it('Source code CVE tree node', () => {
```

**Review by yahavi — CHANGES_REQUESTED** [review]

File: package.json [hunk 30]
```diff
@@ -361,5 +366,11 @@
...
         "artifactory",
         "vulnerabilities"
-    ]
+    ],
+    "__metadata": {
+        "id": "2387d33f-8794-47e8-830e-a0573ba80988",
+        "publisherDisplayName": "JFrog",
+        "publisherId": "38fb619f-642a-42fd-b593-7de38dedfd23",
+        "isPreReleaseVersion": false
+    }
```
Comment (yahavi) [inline package.json:375]: Why is it needed?
```suggestion
```

**Comment by yahavi** [comment]
I got this in DEPENDENCY DETAILS:
image

Intended or not, it looks like a bug

File: src/main/binary/cveApplicabilityRunner.ts [hunk 31]
```diff
@@ -0,0 +1,135 @@
...
+     * @returns Scans a project based on the provided project path. Scans all CVEs or specific ones if 'cveToScan' was provided
+     */
+    public scan(pathToRoot: string, cvesToScan?: string, packageType?: PackageType): string | undefined {
+        try {
+            if (!this._isOsSupported) {
+                return;
+            }
+            let cmdArgs: string = '';
+            if (packageType === PackageType.NPM) {
+                cmdArgs = ' --skipped-folders=node_modules ';
```
[the remark this answers is not in the document]
Comment (Or-Geva) [reply src/main/binary/cveApplicabilityRunner.ts:78]: We cant. VS-code has a different exclusion pattern than applicability scanners.

File: src/main/binary/cveApplicabilityRunner.ts [hunk 32]
```diff
@@ -0,0 +1,135 @@
...
+    private _lastUpdateFile: string;
+    constructor(connectionManager: ConnectionManager, private _logManager: LogManager) {
+        let downloadUrl: string = 'ide-scanners/applicability_scanner/[RELEASE]';
+        let binary: string = 'applicability_scanner';
+        switch (os.platform()) {
+            case 'win32':
+                binary += '.exe';
+                downloadUrl += '/windows/' + binary;
+                break;
+            case 'linux':
```
[the remark this answers is not in the document]
Comment (Or-Geva) [reply src/main/binary/cveApplicabilityRunner.ts:29]: I think it's only Linux on x64 (i.e., no ARM and no x32)

File: src/main/binary/cveApplicabilityRunner.ts [hunk 33]
```diff
@@ -0,0 +1,135 @@
...
+    public async update(): Promise<void> {
+        if (!this.shouldUpdate()) {
+            return;
+        }
+        try {
+            await this._resource.update(true);
+            // Save time when the update accrued.
+            this.saveTime();
+        } catch (error) {
+            this._logManager.logMessage('failed to update the applicable scanner: ' + error, 'WARN');
```
[the remark this answers is not in the document]
Comment (Or-Geva) [reply src/main/binary/cveApplicabilityRunner.ts:60]: I have added a test as well.

File: src/main/diagnostics/applicabilityActionProvider.ts [hunk 34]
```diff
@@ -1,99 +1,103 @@
...
-//                 providedCodeActionKinds: [vscode.CodeActionKind.Empty]
-//             })
-//         );
-//     }
+    activate(context: vscode.ExtensionContext) {
+        this.registerListeners(context.subscriptions);
+        vscode.workspace.textDocuments.forEach(this.updateDiagnostics, this);
+        context.subscriptions.push(
+            this,
+            vscode.languages.registerCodeActionsProvider({ scheme: 'file', pattern: '**/*.{py,js}' }, this, {
```
[the remark this answers is not in the document]
Comment (Or-Geva) [reply src/main/diagnostics/applicabilityActionProvider.ts:21]: Typescript is currently not supported.

File: src/main/scanCache/scanCacheManager.ts [hunk 35]
```diff
@@ -93,75 +76,78 @@ export class ScanCacheManager implements ExtensionComponent {
...
+            for (const [cve, sevirity] of prevObject.cves) {
+                if (scannedCveObject.cves == undefined) {
+                    scannedCveObject.cves = new Map<string, Severity>();
+                }
+                scannedCveObject.cves.set(cve, sevirity);
+            }
+        }
+        fs.writeFileSync(
+            this.getCvesCacheFile(scannedCveObject.projectPath),
+            JSON.stringify(scannedCveObject, (key, value) => {
```
[the remark this answers is not in the document]
Comment (Or-Geva) [reply src/main/scanCache/scanCacheManager.ts:94]: Map objects are not serialized to a file by `JSON.stringify`. As suggested over [here](https://stackoverflow.com/a/56150320/8439478), I followed the steps.

File: src/main/types/applicabilityScanResults.ts (hunk 13 above)
Comment (Or-Geva) [reply src/main/types/applicabilityScanResults.ts:3]: All CVEs that the Applicability scanner searched for (regardless of whether they were found)

File: src/main/utils/resource.ts (hunk 18 above)
Comment (Or-Geva) [reply src/main/utils/resource.ts:13]: This should be overridden for test purposes. The readonly key won't let me set it.

File: src/main/utils/resource.ts (hunk 19 above)
Comment (Or-Geva) [reply src/main/utils/resource.ts:58]: `this._logManager.logMessage('Downloading CVE Applicability scanner from ' + this.downloadTarget, 'DEBUG');`
WDYT?

File: src/main/utils/resource.ts (hunk 21 above)
Comment (Or-Geva) [reply src/main/utils/resource.ts:75]: `this._logManager.logMessage('The '+this.resourceName+' was successfully updated. (' + this.downloadSource+')', 'DEBUG');`
WDYT?

File: src/main/utils/resource.ts (hunk 24 above)
Comment (Or-Geva) [reply src/main/utils/resource.ts:71]: Could you elaborate more on this? What makes windows so important?

File: src/main/utils/resource.ts (hunk 26 above)
Comment (Or-Geva) [reply src/main/utils/resource.ts:85]: The following compile error is occurring:
```
Type '{ sha256: string; }' is missing the following properties from type 'IChecksumResult': sha1, md5
```

File: package.json (hunk 30 above)
Comment (Or-Geva) [reply package.json:375]: Removed

**Comment by srmish-jfrog** [comment]
I've spotted a bug in one of my runs -
When running on https://git.jfrog.info/projects/PIP/repos/controlplane/browse, the CVE-2019-15605 appears under Code Vulnerabilities. However - CVE-2019-15695 is in Node.js itself (not in one of the project dependencies) and indeed doesn't appear under the dependencies tab. So - the contextual analyzer for CVE-2019-15695 should never have been triggered for this project!

image

**Pull request merged**